src/libraries/System.Reflection.DispatchProxy/src/System/Reflection/DispatchProxyGenerator.cs

```diff
@@ -62,16 +62,21 @@ internal static class DispatchProxyGenerator
         private static readonly MethodInfo s_dispatchProxyInvokeMethod = typeof(DispatchProxy).GetTypeInfo().GetDeclaredMethod("Invoke")!;
 
         // Returns a new instance of a proxy the derives from 'baseType' and implements 'interfaceType'
-        internal static object CreateProxyInstance(Type baseType, Type interfaceType)
+        internal static object CreateProxyInstance(
+            [DynamicallyAccessedMembers(DynamicallyAccessedMemberTypes.All)] Type baseType,
+            [DynamicallyAccessedMembers(DynamicallyAccessedMemberTypes.All)] Type interfaceType)
```

Comment:
I'm no expert in this area... but this feels unnecessary. If I were to create the proxy manually by implementing it in source I could only include the members actually used by my application - everything would work. And in that case trimming could remove the unused members from the interface type without breaking anything.

Another point of view is - we could one day create a source generator to basically replace the `CreateProxy` with generated code. If some of the methods generated there were never used by the app, trimming would remove them as well.

So in theory runtime proxy generation is no different - if there are members which are never used on the interface by the app, they don't need proxy implementation - it would not be used either.

I'm also not sure about the base type - I understand that it's because `TypeBuilder` has this requirement already... but it feels weird for the same reasons. In theory there should be no difference in trimming capability between runtime solution using Ref.Emit classes and compile time solution using source generators - but it's probably hard to get there.

Reply:
I'm definitely not an expert in this area either - this was my first encounter with DispatchProxy 😄.

I think what you say makes sense for the interface. If the app isn't using members on the interface, DispatchProxy has no reason to preserve those members. So I think it makes sense removing the DynamicallyAccessedMembers from the interfaceType.

For the `baseType`, we MUST keep `PublicParameterlessConstructor` at a minimum. But for the other members on the `baseType`, the only one needed is the `Invoke` override, which will be preserved because the abstract `Invoke` method on `DispatchProxy` will be preserved (since we get it through Reflection to generate calls to it). So I agree we could relax this annotation from `All` to `PublicParameterlessConstructor`, but at the end of the day is it worth it? It is definitely safer to keep everything on the Type that we are deriving from (just like TypeBuilder is annotated).

@jkotas @stephentoub - Do you have any thoughts here?

Reply:
I agree that it is sufficient to mark this one with `PublicParameterlessConstructor`.

> at the end of the day is it worth it?

Since we have the fine grained annotations, I think we should use where appropriate. I do not see a good reason why we should be conservative here.

> keep everything on the Type that we are deriving from (just like TypeBuilder is annotated).

Do we have the motivation behind this annotation recorded somewhere?
I assume that we are doing this to handle situation where the base type provides implementations by name that the linker has no way of seeing, like:
```
interface IFace
{
    void F();
}

public class Base
{
    public void F() { }
}

// The linker has no way of knowing that it needs to keep Base.F() around?
public class EmittedViaRefEmit : Base, IFace
{
}
```

Reply:
> Do we have the motivation behind this annotation recorded somewhere?

You mean why `TypeBuilder` requires the base type to be annotated as `.All`? It's because the result of type building has `.All` (otherwise reflection emit would be kind of useless). Say I e.g. take this as the base type:

```csharp
class Base
{
    public void Foo() { }
    public int F;
}
```

Then make a `TypeBuilder` that creates a type that derives from this class and then call `GetMethod("Foo")` on the result. Since the result of the `TypeBuilder` was annotated as `.All` this will work without a warning, but we are actually walking the inheritance chain back to the base type (`GetMethod` will walk the inheritance chain if the current type doesn't have method with this name). The only way to make that safe is to require base type to also be annotated as `.All`.

Reply:
Thanks for the help/info. I've made the requested change.